executor/grace_hash_join.go

```diff
@@ -0,0 +1,275 @@
+//                            |              ^
+//                            |              |
+//                            v              |
+//             +------------------------------------------------+
+//             |                  PartitionInDisk               |
+//             +------------------------------------------------+
+//               ^            ^                ^              ^
+//               |            |                |              |
+//               |            |                |              |
+//               v            v                v              v
+//    +------------+    +------------+    +------------+    +------------+
+//    | ListInDisk |    | ListInDisk |    | ListInDisk |    | ListInDisk |
+//    +------------+    +------------+    +------------+    +------------+
+//
+////////////////////////////////////////////////////////////////////////////////////////
+type GraceHashJoinExec struct {
+	baseExecutor
+
+	probeSideExec        Executor
+	buildSideExec        Executor
+	hashJoinExec         Executor
+	buildPartitionerExec *PartitionerExec
+	probePartitionerExec *PartitionerExec
+	buildHashPartitioner *hashPartitioner
+	probeHashPartitioner *hashPartitioner
+
+	probeKeys  []*expression.Column
+	buildKeys  []*expression.Column
+	probeTypes []*types.FieldType
+	buildTypes []*types.FieldType
+
+	memTracker  *memory.Tracker // track memory usage.
+	diskTracker *disk.Tracker   // track disk usage.
+
+	prepared     bool
+	partitionCnt int
+	partitionIdx int
+}
+
+// Open implements the Executor Open interface.
+func (e *GraceHashJoinExec) Open(ctx context.Context) error {
+	if err := e.baseExecutor.Open(ctx); err != nil {
+		return err
+	}
+	if err := e.hashJoinExec.Open(ctx); err != nil {
+		return err
+	}
+
+	e.prepared = false
+	e.memTracker = memory.NewTracker(e.id, -1)
+	e.memTracker.AttachTo(e.ctx.GetSessionVars().StmtCtx.MemTracker)
+
+	e.diskTracker = disk.NewTracker(e.id, -1)
+	e.diskTracker.AttachTo(e.ctx.GetSessionVars().StmtCtx.DiskTracker)
+
+	if e.probeTypes == nil {
+		e.probeTypes = retTypes(e.probeSideExec)
+	}
+	if e.buildTypes == nil {
+		e.buildTypes = retTypes(e.buildSideExec)
+	}
+
+	e.partitionIdx = 0
+
+	buildKeyColIdx := make([]int, len(e.buildKeys))
+	for i := range e.buildKeys {
+		buildKeyColIdx[i] = e.buildKeys[i].Index
+	}
+	buildHashCtx := &hashContext{
+		allTypes:  e.buildTypes,
+		keyColIdx: buildKeyColIdx,
+	}
+	e.buildHashPartitioner = newHashPartitioner(e.ctx, e.partitionCnt, buildHashCtx)
+
+	probeKeyColIdx := make([]int, len(e.probeKeys))
+	for i := range e.probeKeys {
+		probeKeyColIdx[i] = e.probeKeys[i].Index
+	}
+	probeHashCtx := &hashContext{
+		allTypes:  e.probeTypes,
+		keyColIdx: probeKeyColIdx,
+	}
+	e.probeHashPartitioner = newHashPartitioner(e.ctx, e.partitionCnt, probeHashCtx)
+
+	return nil
+}
+
+// Close implements the Executor Close interface.
+func (e *GraceHashJoinExec) Close() error {
+	err := e.buildPartitionerExec.partitionInDisk.Close()
```

Comment:
I think it's better to call `err := e.buildPartitionerExec.Close()` here. And do the `partitionInDisk.Close()` in `PartitionerExec.Close()`. What's your opinion? @b41sh

Reply:
good idea